test/hotspot/jtreg/runtime/Monitor/libMonitorWithDeadObjectTest.c

```diff
@@ -0,0 +1,172 @@
+  if ((*env)->ExceptionCheck(env)) {
+    fprintf(stderr, "Error: %s", msg);
+    exit(-1);
+  }
+}
+
+#define check(env, what, msg)                      \
+  check_exception((env), (msg));                   \
+  do {                                             \
+    if ((what) == 0) {                             \
+      fprintf(stderr, #what "is null: %s", (msg)); \
+      exit(-2);                                    \
+    }                                              \
+  } while (0)
+
+static jobject create_object(JNIEnv* env) {
+  jclass clazz = (*env)->FindClass(env, "java/lang/Object");
+  check(env, clazz, "No class");
+
+  jmethodID constructor = (*env)->GetMethodID(env, clazz, "<init>", "()V");
+  check(env, constructor, "No constructor");
+
+  jobject obj = (*env)->NewObject(env, clazz, constructor);
+  check(env, constructor, "No object");
+
+  return obj;
+}
+
+static void system_gc(JNIEnv* env) {
+  jclass clazz = (*env)->FindClass(env, "java/lang/System");
+  check(env, clazz, "No class");
+
+  jmethodID method = (*env)->GetStaticMethodID(env, clazz, "gc", "()V");
+  check(env, method, "No method");
+
+  (*env)->CallStaticVoidMethod(env, clazz, method);
+  check_exception(env, "Calling System.gc()");
+}
+
+static void thread_dump_with_locked_monitors(JNIEnv* env) {
+  jclass ManagementFactoryClass = (*env)->FindClass(env, "java/lang/management/ManagementFactory");
+  check(env, ManagementFactoryClass, "No ManagementFactory class");
+
+  jmethodID getThreadMXBeanMethod = (*env)->GetStaticMethodID(env, ManagementFactoryClass, "getThreadMXBean", "()Ljava/lang/management/ThreadMXBean;");
+  check(env, getThreadMXBeanMethod, "No getThreadMXBean method");
+
+  jobject threadBean = (*env)->CallStaticObjectMethod(env, ManagementFactoryClass, getThreadMXBeanMethod);
+  check(env, threadBean, "Calling getThreadMXBean()");
+
+  jclass ThreadMXBeanClass = (*env)->FindClass(env, "java/lang/management/ThreadMXBean");
+  check(env, ThreadMXBeanClass, "No ThreadMXBean class");
+
+  jmethodID dumpAllThreadsMethod = (*env)->GetMethodID(env, ThreadMXBeanClass, "dumpAllThreads", "(ZZ)[Ljava/lang/management/ThreadInfo;");
+  check(env, dumpAllThreadsMethod, "No dumpAllThreads method");
+
+  // The 'lockedMonitors == true' is what causes the monitor with a dead object to be examined.
+  jobject array = (*env)->CallObjectMethod(env, threadBean, dumpAllThreadsMethod, JNI_TRUE /* lockedMonitors */, JNI_FALSE /* lockedSynchronizers*/);
+  check(env, array, "Calling dumpAllThreads(true, false)");
+}
+
+static void create_monitor_with_dead_object(JNIEnv* env) {
+  jobject obj = create_object(env);
+
+  if ((*env)->MonitorEnter(env, obj) != 0) die("MonitorEnter");
+
+  // Drop the last strong reference to the object associated with the monitor.
+  // The monitor only keeps a weak reference to the object.
+  (*env)->DeleteLocalRef(env, obj);
+
+  // Let the GC clear the weak reference to the object.
+  system_gc(env);
```

Comment:
A single GC may not be enough...

Reply:
@dcubed-ojdk there's some earlier discussion on this. Apparently a single GC is sufficient to clear an oopStorage WeakHandle, even though it may not be enough for Java level reference processing actions to be observed.

Reply:
One is enough for this test.